go/cpu/bpf/bpf.go

```diff
@@ -0,0 +1,238 @@
+	}
+}
+
+// getJumpElse gets the jump else case offset in bytes
+func (c *BpfCpu) getJumpElse(a arg) uint32 {
+	switch a := a.(type) {
+	case *j:
+		return 0
+	case *jelse:
+		return uint32(a.jt) * 8
+	default:
+		panic("Jump with illegal args!")
+	}
+}
+
+func (c *BpfCpu) Start(begin, until uint64) error {
+	var dis Dis
+	pc := uint32(begin)
+	c.RegWrite(PC, uint64(pc))
+	var err error
+
+	for pc <= uint32(until) && err == nil {
+		var mem []byte
+		var code []models.Ins
+
+		if mem, err = c.ReadProt(uint64(pc), 8, cpu.PROT_EXEC); err != nil {
+			break
+		}
+		if code, err = dis.Dis(mem, uint64(pc)); err != nil {
+			break
+		}
+		ins := code[0].(*ins)
+		fmt.Printf("%04x: %s\n", pc, ins)
+
+		c.OnCode(uint64(pc), uint32(len(ins.bytes)))
+
+		jumpoff := uint32(0)
+		al, _ := c.RegRead(A)
+		a := uint32(al)
+		xl, _ := c.RegRead(X)
+		x := uint32(xl)
+
+		switch ins.name {
+		case "ret":
+			err = fmt.Errorf("Returning value %#x", c.get(ins.arg))
+		case "ld":
+			fallthrough
+		case "ldi":
+			c.RegWrite(A, uint64(c.get(ins.arg)))
+		case "ldh":
+			c.RegWrite(A, uint64(c.get(ins.arg))&0xffff)
+		case "ldb":
+			c.RegWrite(A, uint64(c.get(ins.arg))&0xff)
+		case "ldx":
+			fallthrough
+		case "ldxi":
+			c.RegWrite(X, uint64(c.get(ins.arg)))
+		case "ldxb":
+			c.RegWrite(X, uint64(c.get(ins.arg))&0xff)
+		case "st":
+			c.RegWrite(int(c.get(ins.arg)), uint64(a))
+		case "stx":
+			c.RegWrite(int(c.get(ins.arg)), uint64(x))
+		case "jmp":
+			jumpoff = c.getJump(ins.arg)
+		case "jeq":
+			if c.get(ins.arg) == a {
+				jumpoff = c.getJump(ins.arg)
+			} else {
+				jumpoff = c.getJumpElse(ins.arg)
+			}
+		case "jgt":
+			if c.get(ins.arg) > a {
+				jumpoff = c.getJump(ins.arg)
+			} else {
+				jumpoff = c.getJumpElse(ins.arg)
+			}
+		case "jge":
+			if c.get(ins.arg) >= a {
+				jumpoff = c.getJump(ins.arg)
+			} else {
+				jumpoff = c.getJumpElse(ins.arg)
+			}
+		case "jset":
+			if (c.get(ins.arg) & a) != 0 {
+				jumpoff = c.getJump(ins.arg)
+			} else {
+				jumpoff = c.getJumpElse(ins.arg)
+			}
+		case "add":
+			c.RegWrite(A, uint64(a+c.get(ins.arg)))
+		case "sub":
+			c.RegWrite(A, uint64(a-c.get(ins.arg)))
+		case "mul":
+			c.RegWrite(A, uint64(a*c.get(ins.arg)))
+		case "div":
+			c.RegWrite(A, uint64(a/c.get(ins.arg)))
+		case "mod":
+			c.RegWrite(A, uint64(a%c.get(ins.arg)))
+		case "neg":
+			c.RegWrite(A, uint64(-a))
+		case "and":
+			c.RegWrite(A, uint64(a&c.get(ins.arg)))
+		case "or":
+			c.RegWrite(A, uint64(a&c.get(ins.arg)))
+		case "xor":
+			c.RegWrite(A, uint64(a^c.get(ins.arg)))
+		case "lsh":
+			c.RegWrite(A, uint64(a<<c.get(ins.arg)))
+		case "rsh":
+			c.RegWrite(A, uint64(a>>c.get(ins.arg)))
+		case "tax":
+			c.RegWrite(X, uint64(a))
+		case "txa":
+			c.RegWrite(A, uint64(x))
+		}
+		pc += 8 + jumpoff
+	}
+	return err
+}
+
+// TODO: Implement these
+func (c *BpfCpu) Stop() error {
+	return nil
+}
+
+func (c *BpfCpu) Close() error {
+	return nil
+}
+
+func (c *BpfCpu) Backend() interface{} {
```

Comment:
This one isn't super important, as there's no backend handle for homemade cpus (it's used elsewhere to pull back the curtain and access unicorn internals)